### PR TITLE
plugin host not to crash on activation error and add messages

### PR DESCRIPTION
**Pull request by amiramw**
to notify the end user about plugin activation or loading errors

Signed-off-by: Amiram Wingarten <amiram.wingarten@sap.com>

#### What it does

Fixes a behavior such that exception on activatation method carshes plugin host without showing the exception anywhere which make it impossible to investigate without debugging the plugins.
Now such exception is catched, logged and message is showed to the end user.
Cover some of the issues raised in https://github.com/theia-ide/theia/issues/6061

#### How to test

Create extension with activation event * and throw an exception on the activation method. Make sure that when loading it a proper message is printed to console and showed to the end user. Make sure that plugin host doesn't crash and that other plugins still work.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by akosyakov** [comment]
> Create extension with activation event * and throw an exception on the activation method. Make sure that when loading it a proper message is printed to console and showed to the end user. Make sure that plugin host doesn't crash and that other plugins still work.

@amiramw Could you provide an extension please?

File: packages/plugin-ext/src/plugin/plugin-manager.ts [hunk 1]
```diff
@@ -206,6 +213,7 @@ export class PluginManagerExtImpl implements PluginManagerExt, PluginManager {
                 if (pluginMain !== undefined) {
                     await this.startPlugin(plugin, configStorage, pluginMain);
                 } else {
+                    this.messageRegistryProxy.$showMessage(MainMessageType.Error, `Unable to load a plugin ${plugin.model.name}`, {}, []);
```
[the remark this answers is not in the document]
Comment (akosyakov) [reply packages/plugin-ext/src/plugin/plugin-manager.ts]: it's turned out that if VS Code extension does not have code, then `pluginMain` should be stubbed with `{}`, see https://github.com/TypeFox/vscode/blob/70b8db24a37fafc77247de7f7cb5bb0195120ed0/src/vs/workbench/api/common/extHostExtensionService.ts#L372-L376

cc @theia-ide/plugin-system

**Comment by amiramw** [comment]
@akosyakov here is an example for such problematic extension: https://github.com/amiramw/error-activate-vscode

File: packages/plugin-ext/src/plugin/plugin-manager.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (amiramw) [reply packages/plugin-ext/src/plugin/plugin-manager.ts]: passing undefined does not pass compiler type check

File: packages/plugin-ext/src/plugin/plugin-manager.ts [hunk 2]
```diff
@@ -254,13 +262,18 @@ export class PluginManagerExtImpl implements PluginManagerExt, PluginManager {
...
-            // resolve activation promise
-            if (this.pluginActivationPromises.has(plugin.model.id)) {
-                this.pluginActivationPromises.get(plugin.model.id)!.resolve();
-                this.pluginActivationPromises.delete(plugin.model.id);
+            try {
+                const pluginExport = await pluginMain[plugin.lifecycle.startMethod].apply(getGlobal(), [pluginContext]);
+                this.activatedPlugins.set(plugin.model.id, new ActivatedPlugin(pluginContext, pluginExport, stopFn));
+
+                // resolve activation promise
+                if (this.pluginActivationPromises.has(plugin.model.id)) {
```
[the remark this answers is not in the document]
Comment (amiramw) [reply packages/plugin-ext/src/plugin/plugin-manager.ts]: ok, i will check if it exist and reject it

File: packages/plugin-ext/src/plugin/plugin-manager.ts [hunk 3]
```diff
@@ -196,7 +201,9 @@ export class PluginManagerExtImpl implements PluginManagerExt, PluginManager {
                         if (dependency) {
                             await this.loadPlugin(dependency, configStorage, visited);
                         } else {
-                            console.warn(`cannot find a dependency to '${dependencyId}' for '${plugin.model.id}' plugin`);
+                            const message = `cannot find a dependency to '${dependencyId}' for '${plugin.model.id}' plugin`;
+                            this.messageRegistryProxy.$showMessage(MainMessageType.Warning, message, {}, []);
+                            console.warn(message);
```
[the remark this answers is not in the document]
Comment (amiramw) [reply packages/plugin-ext/src/plugin/plugin-manager.ts]: now the function return boolean for success so i return false in this case

**Comment by amiramw** [comment]
@akosyakov i believe i addressed all the comments. please review.

File: packages/plugin-ext/src/plugin/plugin-manager.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (akosyakov) [reply packages/plugin-ext/src/plugin/plugin-manager.ts]: Do we need a warning in logs anymore?

File: packages/plugin-ext/src/plugin/plugin-manager.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (akosyakov) [reply packages/plugin-ext/src/plugin/plugin-manager.ts]: > passing undefined does not pass compiler type check

Could you share an error? I meant:
```ts
let pluginMain = this.host.loadPlugin(plugin);
// see https://github.com/TypeFox/vscode/blob/70b8db24a37fafc77247de7f7cb5bb0195120ed0/src/vs/workbench/api/common/extHostExtensionService.ts#L372-L376
pluginMain = pluginMain || {};
return this.startPlugin(plugin, configStorage, pluginMain);
```

File: packages/plugin-ext/src/plugin/plugin-manager.ts [hunk 4]
```diff
@@ -255,17 +275,29 @@ export class PluginManagerExtImpl implements PluginManagerExt, PluginManager {
...
+                    this.pluginActivationPromises.get(plugin.model.id)!.reject();
+                }
+                const id = plugin.model.displayName || plugin.model.id;
+                this.messageRegistryProxy.$showMessage(MainMessageType.Error, `Activating extension ${id} failed: ${err.message}.`, {}, []);
+                console.error(`Error on activation of ${plugin.model.name} - ${err}`);
+                return false;
             }
         } else {
-            console.log(`There is no ${plugin.lifecycle.startMethod} method on plugin`);
+            console.log(`There is no ${plugin.lifecycle.startMethod} method on plugin so the module is the extension's exports`);
```
Comment (akosyakov) [inline packages/plugin-ext/src/plugin/plugin-manager.ts]: `startMethod` can be `undefined`, it means that an extension does not have any code, but it still should be activated with `pluginMain` as exports, see: https://github.com/TypeFox/vscode/blob/70b8db24a37fafc77247de7f7cb5bb0195120ed0/src/vs/workbench/api/common/extHostExtensionService.ts#L400-L401

Please also include a link in the code as a comment for future readers.

File: packages/plugin-ext/src/plugin/plugin-manager.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (amiramw) [reply packages/plugin-ext/src/plugin/plugin-manager.ts]: i think so. it allows us to monitor usage of malformed extensions and not rely on end users telling us.

File: packages/plugin-ext/src/plugin/plugin-manager.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (amiramw) [reply packages/plugin-ext/src/plugin/plugin-manager.ts]: i though you meant the $showMessage method. sure i'll change that.

File: packages/plugin-ext/src/plugin/plugin-manager.ts (hunk 4 above)
Comment (amiramw) [reply packages/plugin-ext/src/plugin/plugin-manager.ts]: added the link

File: packages/plugin-ext/src/plugin/plugin-manager.ts [hunk 5]
```diff
@@ -254,18 +269,31 @@ export class PluginManagerExtImpl implements PluginManagerExt, PluginManager {
...
+                this.activatedPlugins.set(plugin.model.id, new ActivatedPlugin(pluginContext, pluginExport, stopFn));
+
+                // resolve activation promise
+                if (this.pluginActivationPromises.has(plugin.model.id)) {
+                    this.pluginActivationPromises.get(plugin.model.id)!.resolve();
+                    this.pluginActivationPromises.delete(plugin.model.id);
+                }
+            } catch (err) {
+                if (this.pluginActivationPromises.has(plugin.model.id)) {
+                    this.pluginActivationPromises.get(plugin.model.id)!.reject();
```
Comment (akosyakov) [inline packages/plugin-ext/src/plugin/plugin-manager.ts]: `reject(err)`

File: packages/plugin-ext/src/plugin/plugin-manager.ts [hunk 6]
```diff
@@ -73,13 +75,15 @@ export class PluginManagerExtImpl implements PluginManagerExt, PluginManager {
 
     private readonly registry = new Map<string, Plugin>();
     private readonly activations = new Map<string, (() => Promise<void>)[] | undefined>();
-    private readonly loadedPlugins = new Map<string, Promise<void>>();
+    // promises to whether loading each plugin has been successful
```
Comment (akosyakov) [inline packages/plugin-ext/src/plugin/plugin-manager.ts]: /** ... */ - for js-doc

**Comment by akosyakov** [comment]
I think it completely resolves https://github.com/theia-ide/theia/issues/6061

**Pull request merged**